cmd/server-mux.go

```diff
@@ -43,64 +51,42 @@ var defaultHTTP1Methods = []string{
 	"CONNECT",
 }
 
-// ConnBuf - contains network buffer to record data
-type ConnBuf struct {
-	buffer []byte
-	unRead bool
-	offset int
-}
-
-// ConnMux - implements a Read() which streams twice the firs bytes from
-// the incoming connection, to help peeking protocol
+// ConnMux - Peeks into the incoming connection for relevant
+// protocol without advancing the underlying net.Conn (io.Reader).
+// ConnMux - allows us to multiplex between TLS and Regular HTTP
+// connections on the same listeners.
 type ConnMux struct {
 	net.Conn
-	lastError error
-	dataBuf   ConnBuf
-}
-
-func longestWord(strings []string) int {
-	maxLen := 0
-	for _, m := range defaultHTTP1Methods {
-		if maxLen < len(m) {
-			maxLen = len(m)
-		}
-	}
-	for _, m := range defaultHTTP2Methods {
-		if maxLen < len(m) {
-			maxLen = len(m)
-		}
-	}
-
-	return maxLen
+	readLen int
+	bufrw   *bufio.ReadWriter
 }
 
 // NewConnMux - creates a new ConnMux instance
-func NewConnMux(c net.Conn) *ConnMux {
-	h1 := longestWord(defaultHTTP1Methods)
-	h2 := longestWord(defaultHTTP2Methods)
-	max := h1
-	if h2 > max {
-		max = h2
+func NewConnMux(c net.Conn, readLen int) *ConnMux {
+	br := bufio.NewReader(c)
+	bw := bufio.NewWriter(c)
+	return &ConnMux{
+		Conn:    c,
+		bufrw:   bufio.NewReadWriter(br, bw),
+		readLen: readLen,
 	}
-	return &ConnMux{Conn: c, dataBuf: ConnBuf{buffer: make([]byte, max+1)}}
 }
 
 // PeekProtocol - reads the first bytes, then checks if it is similar
 // to one of the default http methods
 func (c *ConnMux) PeekProtocol() string {
-	var n int
-	n, c.lastError = c.Conn.Read(c.dataBuf.buffer)
-	if n == 0 || (c.lastError != nil && c.lastError != io.EOF) {
-		return ""
+	buf, err := c.bufrw.Peek(c.readLen)
```

Comment:
global `longestWord` can be used directly here and ConnMux.readLen can be eliminated